File: vi/codes/swift/chapter_stack_and_queue/array_deque.swift
/**
 * File: array_deque.swift
 * Created Time: 2023-02-22
 * Author: nuomi1 (nuomi1@qq.com)
 */

/* Hàng đợi hai đầu dựa trên cách triển khai mảng vòng (circular array) */
class ArrayDeque {
    private var nums: [Int] // Mảng dùng để lưu trữ các phần tử của hàng đợi hai đầu
    private var front: Int // Con trỏ đầu, trỏ đến phần tử ở đầu hàng đợi
    private var _size: Int // Độ dài của hàng đợi hai đầu

    /* Hàm khởi tạo */
    init(capacity: Int) {
        nums = Array(repeating: 0, count: capacity)
        front = 0
        _size = 0
    }

    /* Lấy dung lượng của hàng đợi hai đầu */
    func capacity() -> Int {
        nums.count
    }

    /* Lấy độ dài của hàng đợi hai đầu */
    func size() -> Int {
        _size
    }

    /* Kiểm tra hàng đợi hai đầu có rỗng không */
    func isEmpty() -> Bool {
        size() == 0
    }

    /* Tính chỉ số trong mảng vòng */
    private func index(i: Int) -> Int {
        // Dùng phép chia lấy dư để nối liền đầu và cuối mảng
        // Khi i vượt quá cuối mảng, quay về đầu mảng
        // Khi i vượt quá đầu mảng, quay về cuối mảng
        (i + capacity()) % capacity()
    }

    /* Thêm phần tử vào đầu hàng đợi */
    func pushFirst(num: Int) {
        if size() == capacity() {
            print("Double-ended queue is full")
            return
        }
        // Dùng phép chia lấy dư để đưa front vòng về cuối mảng sau khi vượt quá đầu mảng
        // Thêm num vào đầu hàng đợi
        front = index(i: front - 1)
        // Thêm num vào đầu hàng đợi
        nums[front] = num
        _size += 1
    }

    /* Thêm phần tử vào cuối hàng đợi */
    func pushLast(num: Int) {
        if size() == capacity() {
            print("Double-ended queue is full")
            return
        }
        // Dùng phép chia lấy dư để đưa rear vòng về đầu mảng sau khi vượt quá cuối mảng
        let rear = index(i: front + size())
        // Con trỏ đầu di chuyển lùi một vị trí
        nums[rear] = num
        _size += 1
    }

    /* Xóa phần tử ở đầu hàng đợi */
    func popFirst() -> Int {
        let num = peekFirst()
        // Di chuyển con trỏ đầu lùi một vị trí
        front = index(i: front + 1)
        _size -= 1
        return num
    }

    /* Truy cập phần tử ở cuối hàng đợi */
    func popLast() -> Int {
        let num = peekLast()
        _size -= 1
        return num
    }

    /* Trả về danh sách để in ra */
    func peekFirst() -> Int {
        if isEmpty() {
            fatalError("Deque is empty")
        }
        return nums[front]
    }

    /* Mã điều khiển */
    func peekLast() -> Int {
        if isEmpty() {
            fatalError("Deque is empty")
        }
        // Khởi tạo hàng đợi hai đầu
        let last = index(i: front + size() - 1)
        return nums[last]
    }

    /* Trả về mảng để in ra */
    func toArray() -> [Int] {
        // Thêm phần tử vào hàng đợi
        (front ..< front + size()).map { nums[index(i: $0)] }
    }
}

@main
enum _ArrayDeque {
    /* Mã điều khiển */
    static func main() {
        /* Lấy độ dài của hàng đợi hai đầu */
        let deque = ArrayDeque(capacity: 10)
        deque.pushLast(num: 3)
        deque.pushLast(num: 2)
        deque.pushLast(num: 5)
        print("Deque deque = \(deque.toArray())")

        /* Cập nhật phần tử */
        let peekFirst = deque.peekFirst()
        print("Front element peekFirst = \(peekFirst)")
        let peekLast = deque.peekLast()
        print("Rear element peekLast = \(peekLast)")

        /* Thêm phần tử vào hàng đợi */
        deque.pushLast(num: 4)
        print("After element 4 enqueues at rear, deque = \(deque.toArray())")
        deque.pushFirst(num: 1)
        print("After element 1 enqueues at front, deque = \(deque.toArray())")

        /* Xóa phần tử khỏi hàng đợi */
        let popLast = deque.popLast()
        print("Dequeue rear element = \(popLast), after rear dequeue deque = \(deque.toArray())")
        let popFirst = deque.popFirst()
        print("Dequeue front element = \(popFirst), after front dequeue deque = \(deque.toArray())")

        /* Lấy độ dài của hàng đợi hai đầu */
        let size = deque.size()
        print("Deque length size = \(size)")

        /* Kiểm tra hàng đợi hai đầu có rỗng không */
        let isEmpty = deque.isEmpty()
        print("Is deque empty = \(isEmpty)")
    }
}

File: vi/codes/python/chapter_stack_and_queue/array_deque.py
"""
File: array_deque.py
Created Time: 2023-03-01
Author: krahets (krahets@163.com)
"""


class ArrayDeque:
    """Hàng đợi hai đầu dựa trên cách triển khai mảng vòng"""

    def __init__(self, capacity: int):
        """Hàm khởi tạo"""
        self._nums: list[int] = [0] * capacity
        self._front: int = 0
        self._size: int = 0

    def capacity(self) -> int:
        """Lấy dung lượng của hàng đợi hai đầu"""
        return len(self._nums)

    def size(self) -> int:
        """Lấy độ dài của hàng đợi hai đầu"""
        return self._size

    def is_empty(self) -> bool:
        """Kiểm tra hàng đợi hai đầu có rỗng hay không"""
        return self._size == 0

    def index(self, i: int) -> int:
        """Tính chỉ số mảng vòng"""
        # Sử dụng phép chia lấy dư để nối đầu và cuối mảng lại với nhau
        # Khi i vượt qua cuối mảng, quay trở lại đầu mảng
        # Khi i vượt qua đầu mảng, quay trở lại cuối mảng
        return (i + self.capacity()) % self.capacity()

    def push_first(self, num: int):
        """Thêm phần tử vào đầu hàng đợi"""
        if self._size == self.capacity():
            print("Double-ended queue is full")
            return
        # Con trỏ đầu dịch chuyển sang trái một vị trí
        # Sử dụng phép chia lấy dư để đưa front quay về cuối mảng sau khi vượt qua đầu mảng
        self._front = self.index(self._front - 1)
        # Thêm num vào đầu hàng đợi
        self._nums[self._front] = num
        self._size += 1

    def push_last(self, num: int):
        """Thêm phần tử vào cuối hàng đợi"""
        if self._size == self.capacity():
            print("Double-ended queue is full")
            return
        # Tính con trỏ rear, trỏ đến chỉ số cuối + 1
        rear = self.index(self._front + self._size)
        # Thêm num vào cuối hàng đợi
        self._nums[rear] = num
        self._size += 1

    def pop_first(self) -> int:
        """Xóa phần tử ở đầu hàng đợi"""
        num = self.peek_first()
        # Con trỏ đầu dịch chuyển về sau một vị trí
        self._front = self.index(self._front + 1)
        self._size -= 1
        return num

    def pop_last(self) -> int:
        """Xóa phần tử ở cuối hàng đợi"""
        num = self.peek_last()
        self._size -= 1
        return num

    def peek_first(self) -> int:
        """Truy cập phần tử ở đầu hàng đợi"""
        if self.is_empty():
            raise IndexError("Double-ended queue is empty")
        return self._nums[self._front]

    def peek_last(self) -> int:
        """Truy cập phần tử ở cuối hàng đợi"""
        if self.is_empty():
            raise IndexError("Double-ended queue is empty")
        # Tính chỉ số phần tử cuối
        last = self.index(self._front + self._size - 1)
        return self._nums[last]

    def to_array(self) -> list[int]:
        """Trả về mảng để in"""
        # Chỉ chuyển đổi các phần tử nằm trong phạm vi độ dài hợp lệ
        res = []
        for i in range(self._size):
            res.append(self._nums[self.index(self._front + i)])
        return res


"""Code minh họa"""
if __name__ == "__main__":
    # Khởi tạo hàng đợi hai đầu
    deque = ArrayDeque(10)
    deque.push_last(3)
    deque.push_last(2)
    deque.push_last(5)
    print("double-ended queue deque =", deque.to_array())

    # Truy cập các phần tử
    peek_first: int = deque.peek_first()
    print("Front of the queue element peek_first =", peek_first)
    peek_last: int = deque.peek_last()
    print("Rear of the queue element peek_last =", peek_last)

    # Thêm các phần tử vào hàng đợi
    deque.push_last(4)
    print("Element 4 rear enqueue after deque =", deque.to_array())
    deque.push_first(1)
    print("Element 1 front enqueue after deque =", deque.to_array())

    # Xóa các phần tử khỏi hàng đợi
    pop_last: int = deque.pop_last()
    print("Rear dequeued element =", pop_last, ", rear dequeue after deque =", deque.to_array())
    pop_first: int = deque.pop_first()
    print("Front dequeued element =", pop_first, ", front dequeue after deque =", deque.to_array())

    # Lấy độ dài của hàng đợi hai đầu
    size: int = deque.size()
    print("Length of the double-ended queue size =", size)

    # Kiểm tra hàng đợi hai đầu có rỗng hay không
    is_empty: bool = deque.is_empty()
    print("Is the double-ended queue empty =", is_empty)
